### PR TITLE
ceph-volume: fix simple activate for filestore

### PR DESCRIPTION
**Pull request by guits**
a `simple scan` on a dmcrypt filestore OSD returns the following json:

```
{
    "active": "ok",
    "ceph_fsid": "b295b2c9-241d-4399-a3b2-3dc00b5235f3",
    "cluster_name": "ceph",
    "data": {
        "path": "/dev/sdc1",
        "uuid": "3b46c6cb-528c-4a6c-83d0-b21fd776bcf6"
    },
    "encrypted": true,
    "encryption_type": "luks",
    "fsid": "3b46c6cb-528c-4a6c-83d0-b21fd776bcf6",
    "journal_dmcrypt": {
        "path": "/tmp/tmpPR4Rqc/journal_dmcrypt",
        "uuid": "e27590de-410b-4689-9d02-41ceeb9a0dcc"
    },
    "journal_uuid": "e27590de-410b-4689-9d02-41ceeb9a0dcc",
    "keyring": "AQCJYlxdnsrxHBAAsbZZ66/3qKPS5ywapy8q8A==",
    "lockbox.keyring": "AQCOYlxd9XcbJhAA0+bE4GlKl0GGcTkaJiyV/Q==",
    "magic": "ceph osd volume v026",
    "none": "",
    "ready": "ready",
    "require_osd_release": 14,
    "type": "filestore",
    "whoami": 4
}
```

when dmcrypt, the key `journal.uuid` doesn't exist, it's `journal_dmcrypt.uuid`
instead.

This commit detects whether dmcrypt is enabled in order to use the right
key name (journal vs. journal_dmcrypt)

typical error fixed:

```
+/opt/ceph-container/bin/osd_volume_activate.sh:27: osd_volume_simple(): ceph-volume simple activate --file /etc/ceph/osd/0-59239dbe-ca08-4ae4-b903-330d0b8c8361.json --no-systemd
--> Required devices (data, and journal) not present for filestore
--> filestore devices found: [u'data']
-->  RuntimeError: Unable to activate filestore OSD due to missing devices
```

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

### CODE REVIEW

**Comment by alfredodeza** [comment]
I don't know how this is possible, and it doesn't appear in our tests with ceph-ansible. The output below is from using filestore and dmcrypt with ceph-disk:

```
root@osd0:/etc/ceph/osd# cat 2-dfab9331-33d5-470a-994a-adfc8ef31deb.json 
{
    "active": "ok", 
    "ceph_fsid": "f1c06309-69b3-4169-b487-b12c5a2fba60", 
    "cluster_name": "test", 
    "data": {
        "path": "/dev/sdb1", 
        "uuid": "dfab9331-33d5-470a-994a-adfc8ef31deb"
    }, 
    "encrypted": true, 
    "encryption_type": "plain", 
    "fsid": "dfab9331-33d5-470a-994a-adfc8ef31deb", 
    "journal": {
        "path": "/dev/sdc1", 
        "uuid": "93d0565a-a99d-4e99-b273-ef666d7a7973"
    }, 
    "journal_dmcrypt": {
        "path": "/var/lib/ceph/osd/test-2/journal_dmcrypt", 
        "uuid": "93d0565a-a99d-4e99-b273-ef666d7a7973"
    }, 
    "journal_uuid": "93d0565a-a99d-4e99-b273-ef666d7a7973", 
    "keyring": "AQD3NWBdb1YtJxAA9+zXlTHL6wey6nFuQjt8iA==", 
    "lockbox.keyring": "AQD8NWBdDAjJKxAA2pNr/mbGozF469kqZ7y7/A==", 
    "magic": "ceph osd volume v026", 
    "ready": "ready", 
    "systemd": "", 
    "type": "filestore", 
    "whoami": 2
}root@osd0:/etc/ceph/osd# ls -alh /var/lib/ceph/osd/test-2/
total 56K
drwxr-xr-x 3 ceph ceph  289 Aug 23 21:57 .
drwxr-xr-x 3 ceph ceph 4.0K Aug 23 21:53 ..
-rw-r--r-- 1 root root  180 Aug 23 21:53 activate.monmap
-rw-r--r-- 1 ceph ceph    3 Aug 23 21:53 active
-rw-r--r-- 1 ceph ceph   37 Aug 23 21:53 ceph_fsid
drwxr-xr-x 4 ceph ceph   65 Aug 23 21:53 current
-rw-r--r-- 1 ceph ceph   37 Aug 23 21:53 fsid
lrwxrwxrwx 1 root root   48 Aug 23 21:57 journal -> /dev/mapper/93d0565a-a99d-4e99-b273-ef666d7a7973
lrwxrwxrwx 1 ceph ceph   58 Aug 23 21:53 journal_dmcrypt -> /dev/disk/by-partuuid/93d0565a-a99d-4e99-b273-ef666d7a7973
-rw-r--r-- 1 ceph ceph   37 Aug 23 21:53 journal_uuid
-rw------- 1 ceph ceph   56 Aug 23 21:52 keyring
-rw------- 1 ceph ceph  106 Aug 23 21:55 lockbox.keyring
-rw-r--r-- 1 ceph ceph   21 Aug 23 21:53 magic
-rw-r--r-- 1 ceph ceph    6 Aug 23 21:53 ready
-rw-r--r-- 1 ceph ceph    4 Aug 23 21:53 store_version
-rw-r--r-- 1 ceph ceph   53 Aug 23 21:53 superblock
-rw-r--r-- 1 ceph ceph    0 Aug 23 21:53 systemd
-rw-r--r-- 1 ceph ceph   10 Aug 23 21:53 type
-rw-r--r-- 1 ceph ceph    2 Aug 23 21:53 whoami
```

Which explains why the tests are passing. How is it that an encrypted OSD that has been deployed with ceph-disk is giving you that output? Does it stay that way after a reboot?

**Comment by guits** [comment]
@alfredodeza this only happens in this scenario: filestore/dmcrypt/collocated journal

```ini
[osds]
osd4 osd_objectstore=filestore devices="['/dev/sda', '/dev/sdb', '/dev/sdc']" osd_scenario=collocated
osd5 osd_objectstore=filestore devices="['/dev/sda', '/dev/sdb']" dedicated_devices="['/dev/sdc', '/dev/sdc']" osd_scenario=non-collocated
osd6 osd_objectstore=filestore devices="['/dev/sda', '/dev/sdb', '/dev/sdc']" osd_scenario=collocated dmcrypt=True
osd7 osd_objectstore=filestore devices="['/dev/sda', '/dev/sdb']" dedicated_devices="['/dev/sdc', '/dev/sdc']" osd_scenario=non-collocated dmcrypt=True
```

```bash
[root@osd4]# cat /etc/ceph/osd/14-8df70b89-e1b5-4aa9-9968-4334cce7af2a.json
{
    "active": "ok",
    "ceph_fsid": "b295b2c9-241d-4399-a3b2-3dc00b5235f3",
    "cluster_name": "ceph",
    "data": {
        "path": "/dev/sda1",
        "uuid": "8df70b89-e1b5-4aa9-9968-4334cce7af2a"
    },
    "fsid": "8df70b89-e1b5-4aa9-9968-4334cce7af2a",
    "journal": {
        "path": "/dev/disk/by-partuuid/5eb0e752-6d8f-4a20-bf2c-2828f2f42e28",
        "uuid": "5eb0e752-6d8f-4a20-bf2c-2828f2f42e28"
    },
    "journal_uuid": "5eb0e752-6d8f-4a20-bf2c-2828f2f42e28",
    "keyring": "AQBPY1xdzqMDHhAAZPEOmhLgrIGh2qggqYbAJg==",
    "magic": "ceph osd volume v026",
    "none": "",
    "ready": "ready",
    "type": "filestore",
    "whoami": 14
}
[root@osd4]#
```

```bash
[root@osd5]# cat /etc/ceph/osd/13-cd7a0c3a-9d2e-49cd-adff-47ad29f95611.json
{
    "active": "ok",
    "ceph_fsid": "b295b2c9-241d-4399-a3b2-3dc00b5235f3",
    "cluster_name": "ceph",
    "data": {
        "path": "/dev/sda1",
        "uuid": "cd7a0c3a-9d2e-49cd-adff-47ad29f95611"
    },
    "fsid": "cd7a0c3a-9d2e-49cd-adff-47ad29f95611",
    "journal": {
        "path": "/dev/sdc1",
        "uuid": "b749bf02-3c31-4028-9edb-33bf66ee9453"
    },
    "journal_uuid": "b749bf02-3c31-4028-9edb-33bf66ee9453",
    "keyring": "AQBPY1xdO3nAGRAAXR82gwiYjxnTK9yUhDsmaQ==",
    "magic": "ceph osd volume v026",
    "none": "",
    "ready": "ready",
    "type": "filestore",
    "whoami": 13
}
[root@osd5]#
```

```bash
[root@osd6]# cat /etc/ceph/osd/0-59239dbe-ca08-4ae4-b903-330d0b8c8361.json
{
    "active": "ok",
    "ceph_fsid": "b295b2c9-241d-4399-a3b2-3dc00b5235f3",
    "cluster_name": "ceph",
    "data": {
        "path": "/dev/sda1",
        "uuid": "59239dbe-ca08-4ae4-b903-330d0b8c8361"
    },
    "encrypted": true,
    "encryption_type": "luks",
    "fsid": "59239dbe-ca08-4ae4-b903-330d0b8c8361",
    "journal_dmcrypt": {
        "path": "/tmp/tmpl2IJmf/journal_dmcrypt",
        "uuid": "0d6e085f-95e8-4d58-9e52-3426e6446cf7"
    },
    "journal_uuid": "0d6e085f-95e8-4d58-9e52-3426e6446cf7",
    "keyring": "AQAdYlxdnfbgJxAAtz/GIscpsaonNCZTeiz14A==",
    "lockbox.keyring": "AQAiYlxdIdgHMBAAls0H3Z1DcjanY5MOHgzGzg==",
    "magic": "ceph osd volume v026",
    "none": "",
    "ready": "ready",
    "require_osd_release": 14,
    "type": "filestore",
    "whoami": 0
}
[root@osd6]#
```
```bash
[root@osd7]# cat /etc/ceph/osd/7-6b6fe302-0c0b-4790-8d6f-a07c3a43ecd7.json
{
    "active": "ok",
    "ceph_fsid": "b295b2c9-241d-4399-a3b2-3dc00b5235f3",
    "cluster_name": "ceph",
    "data": {
        "path": "/dev/sda1",
        "uuid": "6b6fe302-0c0b-4790-8d6f-a07c3a43ecd7"
    },
    "encrypted": true,
    "encryption_type": "luks",
    "fsid": "6b6fe302-0c0b-4790-8d6f-a07c3a43ecd7",
    "journal": {
        "path": "/dev/sdc1",
        "uuid": "60d7db4f-a99b-45bd-9de1-3aa6983291a9"
    },
    "journal_dmcrypt": {
        "path": "/tmp/tmpWBiwxU/journal_dmcrypt",
        "uuid": "60d7db4f-a99b-45bd-9de1-3aa6983291a9"
    },
    "journal_uuid": "60d7db4f-a99b-45bd-9de1-3aa6983291a9",
    "keyring": "AQDXYlxdXLI3ORAAdsrcrzJFOnEy9mKenX+y/w==",
    "lockbox.keyring": "AQDdYlxdquHKBRAAoO6TNOln0hewk+MgJt9eAw==",
    "magic": "ceph osd volume v026",
    "none": "",
    "ready": "ready",
    "require_osd_release": 14,
    "type": "filestore",
    "whoami": 7
}
[root@osd7]#
```

in any case, this patch doesn't break other scenarios, IMO it's still worth fixing this.

**Comment by alfredodeza** [comment]
@guits can you check if in the collocated scenario, the OSD is using a file as a journal? if so, then the fix here is to capture the file-based journal with more involvement in the activation part

**Comment by guits** [comment]
@alfredodeza did you get an answer about the value for `allows journal` that seemed to be set differently on your env vs. my env?

**Comment by alfredodeza** [comment]
> @alfredodeza did you get an answer about the value for `allows journal` that seemed to be set differently on your env vs. my env?

If the first OSD is created with bluestore, `allows journal` will say `NO`. If wanting to add OSDs with ceph-disk later, it will mean a file will appear

**Comment by alfredodeza** [comment]
@guits I tried reproducing this again, with a collocated+encrypted ceph-disk OSD and I can't reproduce what you are seeing. You mention that:

> when dmcrypt, the key journal.uuid doesn't exist, it's journal_dmcrypt.uuid

But I don't see any of those:

```
# cat /etc/ceph/osd/1-71e57981-4a96-47ad-8020-27de98bb0b7d.json 
{
    "active": "ok", 
    "ceph_fsid": "c90755e4-9a50-4a2f-909e-f633ebb0c8e0", 
    "cluster_name": "ceph", 
    "data": {
        "path": "/dev/sdac1", 
        "uuid": "71e57981-4a96-47ad-8020-27de98bb0b7d"
    }, 
    "encrypted": true, 
    "encryption_type": "luks", 
    "fsid": "71e57981-4a96-47ad-8020-27de98bb0b7d", 
    "journal": {
        "path": "/dev/sdac2", 
        "uuid": "537d18af-2c5b-4cbb-bf37-8e0c15da2e59"
    }, 
    "journal_dmcrypt": {
        "path": "/var/lib/ceph/osd/ceph-1/journal_dmcrypt", 
        "uuid": "537d18af-2c5b-4cbb-bf37-8e0c15da2e59"
    }, 
    "journal_uuid": "537d18af-2c5b-4cbb-bf37-8e0c15da2e59", 
    "keyring": "AQDEacldD50TFBAAzLOxIMQuVw0P0/UMACncLQ==", 
    "lockbox.keyring": "AQDJacldbU4wGhAAWqJRXFNLUIvn4GmusrutuA==", 
    "magic": "ceph osd volume v026", 
    "ready": "ready", 
    "systemd": "", 
    "type": "filestore", 
    "whoami": 1
}
```
The actual files:

```
# ls -l ceph-1/
total 48
-rw-r--r--. 1 root root 183 Nov 11 14:02 activate.monmap
-rw-r--r--. 1 ceph ceph   3 Nov 11 14:02 active
-rw-r--r--. 1 ceph ceph  37 Nov 11 14:02 ceph_fsid
drwxr-xr-x. 4 ceph ceph  65 Nov 11 14:02 current
-rw-r--r--. 1 ceph ceph  37 Nov 11 14:02 fsid
lrwxrwxrwx. 1 ceph ceph  48 Nov 11 14:02 journal -> /dev/mapper/537d18af-2c5b-4cbb-bf37-8e0c15da2e59
lrwxrwxrwx. 1 ceph ceph  58 Nov 11 14:02 journal_dmcrypt -> /dev/disk/by-partuuid/537d18af-2c5b-4cbb-bf37-8e0c15da2e59
-rw-r--r--. 1 ceph ceph  37 Nov 11 14:02 journal_uuid
-rw-------. 1 ceph ceph  56 Nov 11 14:01 keyring
-rw-r--r--. 1 ceph ceph  21 Nov 11 14:02 magic
-rw-r--r--. 1 ceph ceph   6 Nov 11 14:02 ready
-rw-r--r--. 1 ceph ceph   4 Nov 11 14:02 store_version
-rw-r--r--. 1 ceph ceph  53 Nov 11 14:02 superblock
-rw-r--r--. 1 ceph ceph   0 Nov 11 14:02 systemd
-rw-r--r--. 1 ceph ceph  10 Nov 11 14:02 type
-rw-r--r--. 1 ceph ceph   2 Nov 11 14:02 whoami
```

File: src/ceph-volume/ceph_volume/devices/simple/activate.py [hunk 1]
```diff
@@ -135,6 +136,13 @@ def activate(self, args):
         with open(args.json_config, 'r') as fp:
             osd_metadata = json.load(fp)
 
+        self.is_encrypted = osd_metadata.get('encrypted', False)
```
Comment (alfredodeza) [inline src/ceph-volume/ceph_volume/devices/simple/activate.py]: While trying to reproduce, both `journal` and `journal_dmcrypt` are present in a dmcrypt OSD deployed with ceph-disk:

```
# cat /etc/ceph/osd/1-71e57981-4a96-47ad-8020-27de98bb0b7d.json 
{
    "active": "ok", 
    "ceph_fsid": "c90755e4-9a50-4a2f-909e-f633ebb0c8e0", 
    "cluster_name": "ceph", 
    "data": {
        "path": "/dev/sdac1", 
        "uuid": "71e57981-4a96-47ad-8020-27de98bb0b7d"
    }, 
    "encrypted": true, 
    "encryption_type": "luks", 
    "fsid": "71e57981-4a96-47ad-8020-27de98bb0b7d", 
    "journal": {
        "path": "/dev/sdac2", 
        "uuid": "537d18af-2c5b-4cbb-bf37-8e0c15da2e59"
    }, 
    "journal_dmcrypt": {
        "path": "/var/lib/ceph/osd/ceph-1/journal_dmcrypt", 
        "uuid": "537d18af-2c5b-4cbb-bf37-8e0c15da2e59"
    }, 
    "journal_uuid": "537d18af-2c5b-4cbb-bf37-8e0c15da2e59", 
    "keyring": "AQDEacldD50TFBAAzLOxIMQuVw0P0/UMACncLQ==", 
    "lockbox.keyring": "AQDJacldbU4wGhAAWqJRXFNLUIvn4GmusrutuA==", 
    "magic": "ceph osd volume v026", 
    "ready": "ready", 
    "systemd": "", 
    "type": "filestore", 
    "whoami": 1
}
```

**Comment by alfredodeza** [comment]
Since I can't reproduce this by running ceph-disk, I would love to find a reproducer for this, and if at all possible, how to get there by running ceph-disk directly (not ceph-ansible+container).

This is how I ran ceph-disk to produce a collocated filestore OSD:

```
$ sudo ceph-disk prepare --filestore --dmcrypt /dev/sdac
```

**Pull request closed without merging**